src/ai/backend/storage/filebrowser/filebrowser.py

```diff
@@ -0,0 +1,190 @@
+
+import asyncio
+import logging
+from datetime import datetime
+
+import aiodocker
+
+from ai.backend.common.logging import BraceStyleAdapter
+from ai.backend.common.validators import BinarySize
+from ai.backend.storage.context import Context
+from ai.backend.storage.utils import (
+    get_available_port,
+    is_port_in_use,
+    mangle_path,
+)
+
+from .config_browser_app import prepare_filebrowser_app_config
+from .database import SQLite_DB
+
+log = BraceStyleAdapter(logging.getLogger(__name__))
+
+__all__ = (
+    "create_or_update",
+    "destroy_container",
+    "get_container_by_id",
+    "get_filebrowsers",
+    "get_network_stats",
+)
+
+
+async def create_or_update(ctx: Context, vfolders: list[dict]) -> tuple[str, int, str]:
+    image = ctx.local_config["filebrowser"]["image"]
+    service_ip = ctx.local_config["filebrowser"]["service-ip"]
+    service_port = ctx.local_config["filebrowser"]["service_port"]
+    max_containers = ctx.local_config["filebrowser"]["max-containers"]
+    cgroup = ctx.local_config["filebrowser"]["cgroup"]
+    settings_path = ctx.local_config["filebrowser"]["settings_path"]
+    mount_path = ctx.local_config["filebrowser"]["mount_path"]
+    cpu_count = ctx.local_config["filebrowser"]["max-cpu"]
+    memory = ctx.local_config["filebrowser"]["max-mem"]
+    memory = int(BinarySize().check_and_return(memory))
+    db_path = ctx.local_config["filebrowser"]["db-path"]
+
+    if is_port_in_use(service_port):
+        service_port = get_available_port()
+    running_docker_containers = await get_filebrowsers()
+    if len(running_docker_containers) >= max_containers:
+        print(
+            "Can't create new container. Number of containers exceed the maximum limit.",
+        )
+        return ("0", 0, "0")
+
+    await prepare_filebrowser_app_config(settings_path, service_port)
+
+    docker = aiodocker.Docker()
+
+    config = {
+        "Cmd": [
+            "/filebrowser_dir/start.sh",
+            f"{service_port}",
+        ],
+        "ExposedPorts": {
+            f"{service_port}/tcp": {},
+        },
+        "Image": image,
+        "HostConfig": {
+            "PortBindings": {
+                f"{service_port}/tcp": [
+                    {
+                        "HostIp": f"{service_ip}",
+                        "HostPort": f"{service_port}/tcp",
+                    },
+                ],
+            },
+            "CpuCount": cpu_count,
+            "Memory": memory,
+            "Cgroup": cgroup,
+            "Mounts": [
+                {
+                    "Target": "/filebrowser_dir/",
+                    "Source": f"{settings_path}",
+                    "Type": "bind",
+                },
+            ],
+        },
+    }
+    for vfolder in vfolders:
+        config["HostConfig"]["Mounts"].append(
+            {
+                "Target": f"/data/{str(vfolder['name'])}",
+                "Source": f"{str(mangle_path(mount_path, vfolder['vfid']))}",
```

Comment:
Could we make this code to use `abc.AbstractVolume.mangle_path()` by taking an instance of volume to single-source the method?
It will get a major change in the future version for the vfolder strucuture v3 and I'd like to avoid missing updates due to duplication.